### PR TITLE
Add slightly more structure to the logging in the `CreateSubscriptionController`

### PR DESCRIPTION
**Pull request by jamesgorrie**
## What are you doing in this PR?
Adding more structure and hopefully less noise to the `CreateSubscriptionController`.

This is by no means a silver bullet but a spike to explore and surface the reasons debugging is a bit fiddly, which we could potentially scale out if it feels like a good idea.

[**Trello Card**](https://trello.com)

## Why are you doing this?
Over the past month, nearly all of our alerts we run through have come from ["URGENT 9-5 - PROD support-frontend create recurring product call failed" (`ServerSideCreateFailureAlarm`) alarm](https://github.com/guardian/support-frontend/blob/c4971320c1bbd84d6be4df27990ab8b60cd8690b/cdk/lib/frontend.ts#L295-L318)

While debugging there have been a few issues that make the process a little fiddly. This PR is trying to address filtering through the noise of our logs.

There is quite a lot of logs that occur, mostly useful. But it can make working out the code path that created an failure hard to follow. This PR introduces a structure to the log that we could use to filter out a code path:

`[$Controller] [${request.uuid}] [${userDesc._1}] [${userDesc._2}] [${request.body.product.describe}] $message`

Another noise reduction strategy is to remove the debug info from this log path and to only log the data creating the request. [This data is still logged in the `CustomActionBuilder`](https://github.com/guardian/support-frontend/blob/main/support-frontend/app/actions/CustomActionBuilders.scala#L85).

I know there are more scalable and reusable ways to do this like [`Markers`](https://www.playframework.com/documentation/2.9.x/ScalaLogging#Using-Markers-and-Marker-Contexts), but I really want to validate that this will help before creating the potentially wrong abstraction around it. 

### Failed

```logs
2023-11-24T16:34:15,845+0000 [application-akka.actor.default-dispatcher-8] INFO  com.gu.monitoring.SafeLogger$ - [757ecfff-f824-b6a8-0000-000000000044] [james.gorrie@guardian.co.uk] [guest] [Monthly-SupporterPlus-GBP-10] attempting to create
2023-11-24T16:34:15,854+0000 [application-akka.actor.default-dispatcher-8] INFO  com.gu.monitoring.SafeLogger$ - [757ecfff-f824-b6a8-0000-000000000044] [james.gorrie@guardian.co.uk] [guest] [Monthly-SupporterPlus-GBP-10] {
  "title" : null,
  "firstName" : "James",
  "lastName" : "Gorrie",
  "billingAddress" : {
    "lineOne" : null,
    "lineTwo" : null,
    "city" : null,
    "state" : null,
    "postCode" : null,
    "country" : "GB"
  },
  "deliveryAddress" : null,
  "giftRecipient" : null,
  "product" : {
    "amount" : 10,
    "currency" : "GBP",
    "billingPeriod" : "Monthly",
    "productType" : "SupporterPlus"
  },
  "firstDeliveryDate" : null,
  "paymentFields" : {
    "accountHolderName" : "Whatever",
    "sortCode" : "200000",
    "accountNumber" : "00000000",
    "recaptchaToken" : "***"
  },
  "promoCode" : null,
  "csrUsername" : null,
  "salesforceCaseId" : null,
  "ophanIds" : {
    "pageviewId" : "***",
    "browserId" : null
  },
  "referrerAcquisitionData" : {
    "campaignCode" : null,
    "referrerPageviewId" : "",
    "referrerUrl" : null,
    "componentId" : null,
    "componentType" : null,
    "source" : null,
    "abTests" : null,
    "queryParameters" : [
    ],
    "hostname" : null,
    "gaClientId" : null,
    "userAgent" : null,
    "ipAddress" : null,
    "labels" : null
  },
  "supportAbTests" : [
    {
      "name" : "makeItAnnualNudgeGlobal",
      "variant" : "variant"
    }
  ],
  "email" : "james.gorrie@guardian.co.uk",
  "telephoneNumber" : null,
  "deliveryInstructions" : null,
  "debugInfo" : null
}
2023-11-24T16:34:15,994+0000 [application-akka.actor.default-dispatcher-8] ERROR com.gu.monitoring.SafeLogger$ - [757ecfff-f824-b6a8-0000-000000000044] [james.gorrie@guardian.co.uk] [guest] [Monthly-SupporterPlus-GBP-10] recaptcha failed
2023-11-24T16:34:15,995+0000 [application-akka.actor.default-dispatcher-8] ERROR com.gu.monitoring.SafeLogger$ - [757ecfff-f824-b6a8-0000-000000000044] [james.gorrie@guardian.co.uk] [guest] [Monthly-SupporterPlus-GBP-10] create failed due to RequestValidationError(Recaptcha validation failed)

```

### Succeeded
```
2023-11-24T16:42:12,547+0000 [application-akka.actor.default-dispatcher-9] INFO  com.gu.monitoring.SafeLogger$ - [CreateSubscriptionController] [65360787-7809-9c4c-0000-000000000047] [james.gorrie@guardian.co.uk] [guest] [Monthly-SupporterPlus-GBP-10] attempting to create
2023-11-24T16:42:12,564+0000 [application-akka.actor.default-dispatcher-9] INFO  com.gu.monitoring.SafeLogger$ - [CreateSubscriptionController] [65360787-7809-9c4c-0000-000000000047] [james.gorrie@guardian.co.uk] [guest] [Monthly-SupporterPlus-GBP-10] {
  "title" : null,
  "firstName" : "James",
  "lastName" : "Gorrie",
  "billingAddress" : {
    "lineOne" : null,
    "lineTwo" : null,
    "city" : null,
    "state" : null,
    "postCode" : null,
    "country" : "GB"
  },
  "deliveryAddress" : null,
  "giftRecipient" : null,
  "product" : {
    "amount" : 10,
    "currency" : "GBP",
    "billingPeriod" : "Monthly",
    "productType" : "SupporterPlus"
  },
  "firstDeliveryDate" : null,
  "paymentFields" : {
    "accountHolderName" : "Whatever",
    "sortCode" : "200000",
    "accountNumber" : "00000000",
    "recaptchaToken" : "***"
  },
  "promoCode" : null,
  "csrUsername" : null,
  "salesforceCaseId" : null,
  "ophanIds" : {
    "pageviewId" : "***",
    "browserId" : null
  },
  "referrerAcquisitionData" : {
    "campaignCode" : null,
    "referrerPageviewId" : "",
    "referrerUrl" : null,
    "componentId" : null,
    "componentType" : null,
    "source" : null,
    "abTests" : null,
    "queryParameters" : [
    ],
    "hostname" : null,
    "gaClientId" : null,
    "userAgent" : null,
    "ipAddress" : null,
    "labels" : null
  },
  "supportAbTests" : [
    {
      "name" : "makeItAnnualNudgeGlobal",
      "variant" : "variant"
    }
  ],
  "email" : "james.gorrie@guardian.co.uk",
  "telephoneNumber" : null,
  "deliveryInstructions" : null,
  "debugInfo" : null
}
2023-11-24T16:42:12,713+0000 [application-akka.actor.default-dispatcher-9] INFO  com.gu.monitoring.SafeLogger$ - [CreateSubscriptionController] [65360787-7809-9c4c-0000-000000000047] [james.gorrie@guardian.co.uk] [guest] [Monthly-SupporterPlus-GBP-10] createSubscription
2023-11-24T16:42:13,762+0000 [application-akka.actor.default-dispatcher-9] INFO  com.gu.monitoring.SafeLogger$ - [CreateSubscriptionController] [65360787-7809-9c4c-0000-000000000047] [james.gorrie@guardian.co.uk] [guest] [Monthly-SupporterPlus-GBP-10] create succeeded
```

### CODE REVIEW

File: support-frontend/app/services/stepfunctions/SupportWorkersClient.scala [hunk 1]
```diff
@@ -33,6 +33,7 @@ object CreateSupportWorkersRequest {
   implicit val giftRecipientCodec: Codec[GiftRecipientRequest] = deriveCodec
 
   implicit val decoder: Decoder[CreateSupportWorkersRequest] = deriveDecoder[CreateSupportWorkersRequest]
+  implicit val encoder: Encoder[CreateSupportWorkersRequest] = deriveEncoder[CreateSupportWorkersRequest]
```
Comment (johnduffell) [inline support-frontend/app/services/stepfunctions/SupportWorkersClient.scala:36]: you should be able to use Codec if you need both a Decoder and Encoder

File: support-frontend/app/services/RecaptchaService.scala [hunk 2]
```diff
@@ -26,6 +28,9 @@ class RecaptchaService(wsClient: WSClient)(implicit ec: ExecutionContext) {
       .withMethod("POST")
       .execute()
       .attemptT
-      .leftMap(_.toString)
+      .leftMap(err => {
+        SafeLogger.error(scrub"Recaptcha failed on ${err.toString}")
+        err.toString
+      })
```
Comment (johnduffell) [inline support-frontend/app/services/RecaptchaService.scala:34]: another option to save on brackets (I don't have a strong preference but it's what I normally do)
```suggestion
      .leftMap { err => 
        SafeLogger.error(scrub"Recaptcha failed on ${err.toString}")
        err.toString
      }
```

File: support-frontend/app/services/stepfunctions/SupportWorkersClient.scala (hunk 1 above)
Comment (jamesgorrie) [reply support-frontend/app/services/stepfunctions/SupportWorkersClient.scala]: Good shout

File: support-frontend/app/controllers/CreateSubscriptionController.scala [hunk 3]
```diff
@@ -217,13 +242,11 @@ class CreateSubscriptionController(
       result: EitherT[Future, CreateSubscriptionError, StatusResponse],
       product: ProductType,
       userEmail: String,
-  )(implicit request: Request[CreateSupportWorkersRequest], writeable: Writeable[String]): Future[Result] = {
+  )(implicit request: CreateRequest, writeable: Writeable[String]): Future[Result] = {
```
Comment (johnduffell) [inline support-frontend/app/controllers/CreateSubscriptionController.scala:245]: I know you didn't add it, but I'm not sure why request is implicit, and I prefer to avoid using them unless it saves a lot of boilerplate.  Not sure if it's something for this PR though, it's certainly not a show stopper!

File: support-frontend/app/controllers/CreateSubscriptionController.scala [hunk 4]
```diff
@@ -150,15 +160,30 @@ class CreateSubscriptionController(
...
+    val maybeLoggedInIdentityIdAndEmail =
+      request.user.map(authIdUser => IdentityIdAndEmail(authIdUser.id, authIdUser.primaryEmailAddress))
+
+    val userDesc = maybeLoggedInIdentityIdAndEmail match {
+      case None => (request.body.email, "guest")
+      case Some(idAndEmail) => (idAndEmail.primaryEmailAddress, "logged-in")
+    }
+    SafeLogger.info(
+      s"[CreateSubscriptionController] [${request.uuid}] [${userDesc._1}] [${userDesc._2}] [${request.body.product.describe}] $message",
+    )
```
Comment (johnduffell) [inline support-frontend/app/controllers/CreateSubscriptionController.scala:186]: I usually avoid _1 and _2
```suggestion
    val (user, desc) = maybeLoggedInIdentityIdAndEmail match {
      case None => (request.body.email, "guest")
      case Some(idAndEmail) => (idAndEmail.primaryEmailAddress, "logged-in")
    }
    SafeLogger.info(
      s"[CreateSubscriptionController] [${request.uuid}] [$user] [$desc] [${request.body.product.describe}] $message",
    )
```

File: support-frontend/app/controllers/CreateSubscriptionController.scala [hunk 5]
```diff
@@ -150,15 +160,30 @@ class CreateSubscriptionController(
         }
     }
 
-  private def logIncomingRequest(
-      request: OptionalAuthRequest[CreateSupportWorkersRequest],
-      maybeLoggedInIdentityIdAndEmail: Option[IdentityIdAndEmail],
-  ) = {
+  private def errorDetailedMessage(message: String)(implicit request: CreateRequest) = {
```
Comment (johnduffell) [inline support-frontend/app/controllers/CreateSubscriptionController.scala:163]: these two methods are the same apart from error vs  info?  If so consider to just make one method
e.g. one that takes a `String => Unit` as a parameter and pass in either SafeLogger.error or SafeLogger.info as the parameter

Comment (jamesgorrie) [reply support-frontend/app/controllers/CreateSubscriptionController.scala:163]: I thought about it - but it felt clearer and easier to change in future to just copy/paste until I had another usecase.

File: support-frontend/app/controllers/CreateSubscriptionController.scala [hunk 6]
```diff
@@ -106,22 +112,26 @@ class CreateSubscriptionController(
...
         .leftMap(err => RequestValidationError(err))
-        .map(_.success)
         .subflatMap {
-          case true => Right(())
-          case false => Left(RequestValidationError("Recaptcha validation failed"))
+          case RecaptchaResponse(true, _) => Right(())
+
+          case RecaptchaResponse(false, _) =>
+            errorDetailedMessage("recaptcha failed")
+            Left(RequestValidationError("Recaptcha validation failed"))
```
Comment (johnduffell) [inline support-frontend/app/controllers/CreateSubscriptionController.scala:120]: is this change just for clarity as I can't work out any functionality change?

File: support-frontend/app/controllers/CreateSubscriptionController.scala [hunk 7]
```diff
@@ -61,37 +61,43 @@ class CreateSubscriptionController(
...
+  private type CreateRequest = OptionalAuthRequest[CreateSupportWorkersRequest]
+
   def create: EssentialAction =
     LoggingAndAlarmOnFailure {
       MaybeAuthenticatedActionOnFormSubmission.async(circe.json[CreateSupportWorkersRequest]) { implicit request =>
         implicit val settings: AllSettings = settingsProvider.getAllSettings()
-        SafeLogger.info(s"${request.uuid}: debug info ${request.body.debugInfo}")
+
+        logDetailedMessage("attempting to create")
+        val json = request.body.copy(debugInfo = None).asJson
```
Comment (johnduffell) [inline support-frontend/app/controllers/CreateSubscriptionController.scala:73]: are we stopping logging the debugInfo because it has already been logged in the raw incming request?

Comment (jamesgorrie) [reply support-frontend/app/controllers/CreateSubscriptionController.scala:73]: Indeed. From the description:

Another noise reduction strategy is to remove the debug info from this log path and to only log the data creating the request. [This data is still logged in the `CustomActionBuilder`](https://github.com/guardian/support-frontend/blob/main/support-frontend/app/actions/CustomActionBuilders.scala#L85) (should have dropped it here as a comment).

We've found it makes debugging very hard. @GHaberis suggested we might have had this for when we were trying to reduce a large amount of alarms, where at the moment we only get a few and rarely look at this info.

**Review by johnduffell — APPROVED** [review]
great work 👍 I like logging!

File: support-frontend/app/controllers/CreateSubscriptionController.scala (hunk 6 above)
Comment (jamesgorrie) [reply support-frontend/app/controllers/CreateSubscriptionController.scala:120]: It was actually to allow us to access the `score` values of the `RecaptchaResponse`, which I think we're deemed we don't need. Happy to revert out.

File: support-frontend/app/controllers/CreateSubscriptionController.scala (hunk 3 above)
Comment (jamesgorrie) [reply support-frontend/app/controllers/CreateSubscriptionController.scala:245]: Yea, happy to relook at the clarity of the code in this file after if we want.

File: support-frontend/app/controllers/CreateSubscriptionController.scala (hunk 7 above)
Comment (johnduffell) [reply support-frontend/app/controllers/CreateSubscriptionController.scala:73]: yeah this was hacked in because we had no clue how to reproduce bad requests where e.g. someone selected the payment method first, then went back and filled in the delivdry county, and then the billing address...etc
It was kind of a poor version of quantum metric (although we can include PII in it) to get the redux history in the logs.
If quantum metric covers everything we need then all the more reason to retire it.
I know there were some issues where state changes that went via react useState got missed in the debug info, so it may be less useful anyway.

Comment (johnduffell) [reply support-frontend/app/controllers/CreateSubscriptionController.scala:73]: and yes self-comments are good 💯 , I tend to effectively self-review my PRs because then it keeps the explanation in context (and can help github archaeologists!)
